### DISCUSSION TITLE
Can we run Kafka as Source for Morpheus AutoEncoder pipeline ?

### DISCUSSION THREAD

**Discussion by srisudha-alt** [nv-morpheus/Morpheus#1826] [Experimental] [answered]
<img width="761" alt="ae-pipeline" src="https://github.com/user-attachments/assets/fa5f745d-4170-4301-b34e-0d4b1b16cd5c">
The above documentation shows that AE pipeline don't have support for Kafka as source. 

My Question is "Do we still don't have support for Kafka in AE pipeline ? " 

Thank you in advance !

**Comment by efajardo-nv** [comment]
@srisudha-alt That command is for [Starter DFP](https://github.com/nv-morpheus/Morpheus/tree/branch-24.10/examples/digital_fingerprinting/starter) which will be removed soon in favor of [Production DFP](https://github.com/nv-morpheus/Morpheus/tree/branch-24.10/examples/digital_fingerprinting/production). Here is the current [issue](https://github.com/nv-morpheus/Morpheus/issues/1715) for its removal. It's a stale implementation so we strongly encourage everyone to use Production DFP.

Here's more documentation on Production DFP:
https://docs.nvidia.com/morpheus/developer_guide/guides/5_digital_fingerprinting.html
https://docs.nvidia.com/morpheus/developer_guide/guides/6_digital_fingerprinting_reference.html
https://docs.nvidia.com/morpheus/developer_guide/guides/10_modular_pipeline_digital_fingerprinting.html

Morpheus makes it easy to swap out sources/stages so you should be able update our examples to use the [KafkaSourceStage](https://docs.nvidia.com/morpheus/_modules/morpheus.stages.input.kafka_source_stage.kafkasourcestage.html#morpheus.stages.input.kafka_source_stage.KafkaSourceStage).

**Comment by srisudha-alt** [comment]
@efajardo-nv  Thanks for the links! They are being useful.

I am trying to build a production pipeline using Kafka as Source.
A simple flow like reading from Kafka and writing to a file worked fine. 
(KafkaSourceStage => DeserializeStage => SerializeStage => WriteToFileStage )

Now I am trying to build a full length pipeline where we are trying to do anomaly detection. Pipeline looks like below :
KafkaSourceStage => DFPFileBatcherStage => DFPFileToDataFrameStage = > DFPSplitUsersStage => --------------------------

I got struck at the very first step itself where facing below error which basically shows some incompatibility in outputs and inputs for stage 1 and 2. Can i get some help here ? I couldn't find examples of doing this flow.

====Pipeline Pre-build====
====Pre-Building Segment: linear_segment_0====
Error occurred during Pipeline.build(). Exiting.
Traceback (most recent call last):
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/morpheus/pipeline/pipeline.py", line 458, in build_and_start
    self.build()
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/morpheus/pipeline/pipeline.py", line 294, in build
    self._pre_build()
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/morpheus/pipeline/pipeline.py", line 257, in _pre_build
    stage._pre_build()
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/morpheus/pipeline/source_stage.py", line 73, in _pre_build
    return super()._pre_build(do_propagate=do_propagate)
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/morpheus/pipeline/stage_base.py", line 374, in _pre_build
    dep._pre_build(do_propagate=do_propagate)
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/morpheus/pipeline/stage_base.py", line 351, in _pre_build
    self._pre_compute_schema(schema)
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/morpheus/pipeline/single_port_stage.py", line 65, in _pre_compute_schema
    raise RuntimeError((f"The {self.name} stage cannot handle input of {input_type}. "
RuntimeError: The dfp-file-batcher stage cannot handle input of <class 'morpheus.messages.message_meta.MessageMeta'>. Accepted input types: (<class 'fsspec.core.OpenFiles'>,)
====Pipeline Complete====

**Comment by efajardo-nv** [comment]
@srisudha-alt You are correct. The issue you're seeing is incompatibility of stage 1 output type with the stage 2 input type. The Production DFP example [here](https://github.com/nv-morpheus/Morpheus/blob/branch-24.06/examples/digital_fingerprinting/production/README.md) uses the [MultiFileSource](https://github.com/nv-morpheus/Morpheus/blob/branch-24.06/examples/digital_fingerprinting/production/morpheus/dfp/stages/multi_file_source.py) stage which outputs a list of files (`fsspec.core.OpenFiles`) e.g. from local file system or S3. This is the expected input of [DFPFileBatcherStage](https://github.com/nv-morpheus/Morpheus/blob/branch-24.06/examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_file_batcher_stage.py). The `KafkaSourceStage` emits `MessageMeta` which is not supported by DFPFileBatcherStage and the reason why you see that error.

We have another Production DFP example that builds a similar pipeline with Morpheus modules (instead of pure stages as above) and uses a Kafka source. More information on modules and how to run the examples can be found [here](https://github.com/nv-morpheus/Morpheus/blob/branch-24.06/docs/source/developer_guide/guides/10_modular_pipeline_digital_fingerprinting.md). Module-based pipelines use [control messages](https://github.com/nv-morpheus/Morpheus/blob/branch-24.06/docs/source/developer_guide/guides/9_control_messages.md) which is the expected input of the [file batcher module](https://github.com/nv-morpheus/Morpheus/blob/branch-24.06/morpheus/modules/file_batcher.py). Please look at [dfp_integrated_training_streaming_pipeline.py](https://github.com/nv-morpheus/Morpheus/blob/branch-24.06/examples/digital_fingerprinting/production/morpheus/dfp_integrated_training_streaming_pipeline.py) to see how the example pipeline is constructed.

**Comment by srisudha-alt** [comment]
@efajardo-nv Thank you for your inputs. I am trying to look into and run the intergrated pipelines. I am able to run [dfp_integrated_training_batch_pipeline](https://github.com/nv-morpheus/Morpheus/blob/branch-24.06/examples/digital_fingerprinting/production/morpheus/dfp_integrated_training_batch_pipeline.py) and understood the flow of module based pipelines.
I have tried running running [dfp_integrated_training_streaming_pipeline.py](https://github.com/nv-morpheus/Morpheus/blob/branch-24.06/examples/digital_fingerprinting/production/morpheus/dfp_integrated_training_streaming_pipeline.py) giving [azure_streaming_lti.json](https://github.com/nv-morpheus/Morpheus/blob/branch-24.06/examples/digital_fingerprinting/production/morpheus/control_messages/azure_streaming_lti.json) as input data using a Kafka Topic. But I am facing below issue:

FileToDF [training_pipe]: 0 messages [00:07, ? messages/s]
Error converting payload to ControlMessage : Expecting property name enclosed in double quotes: line 1 column 2 (char 1)
Preprocessed [training_pipe]: 0 messages [00:07, ? messages/s]
Error converting payload to ControlMessage : Extra data: line 1 column 10 (char 9)
Train & Saved [training_pipe]: 0 messages [00:07, ? messages/s]
Error converting payload to ControlMessage : Expecting property name enclosed in double quotes: line 1 column 5 (char 4)
SplitUsers [inference_pipe]: 0 messages [00:07, ? messages/s]
Error converting payload to ControlMessage : Extra data: line 1 column 10 (char 9)
Inferenced [inference_pipe]: 0 messages [00:06, ? messages/s]
Error converting payload to ControlMessage : Expecting property name enclosed in double quotes: line 1 column 6 (char 5)

Error converting payload to ControlMessage : Extra data: line 1 column 10 (char 9)

Error converting payload to ControlMessage : Extra data: line 1 column 16 (char 15)

Error converting payload to ControlMessage : Extra data: line 1 column 17 (char 16)

Error converting payload to ControlMessage : Extra data: line 1 column 13 (char 12)

Error converting payload to ControlMessage : 'str' object has no attribute 'get'

Error converting payload to ControlMessage : Expecting value: line 1 column 6 (char 5)

Error converting payload to ControlMessage : Expecting value: line 1 column 4 (char 3)

Error converting payload to ControlMessage : Expecting value: line 1 column 5 (char 4)

Error converting payload to ControlMessage : Expecting property name enclosed in double quotes: line 1 column 6 (char 5)

Error converting payload to ControlMessage : Extra data: line 1 column 10 (char 9)

Error converting payload to ControlMessage : Extra data: line 1 column 16 (char 15)

Error converting payload to ControlMessage : Expecting value: line 1 column 4 (char 3)

Error converting payload to ControlMessage : Expecting value: line 1 column 5 (char 4)

Error converting payload to ControlMessage : Expecting value: line 1 column 3 (char 2)

Error converting payload to ControlMessage : Extra data: line 1 column 13 (char 12)

Error converting payload to ControlMessage : Extra data: line 1 column 16 (char 15)

Error converting payload to ControlMessage : Extra data: line 1 column 23 (char 22)

Error converting payload to ControlMessage : Extra data: line 1 column 25 (char 24)

Error converting payload to ControlMessage : Extra data: line 1 column 20 (char 19)

Error converting payload to ControlMessage : Extra data: line 1 column 13 (char 12)

Error converting payload to ControlMessage : Expecting value: line 1 column 5 (char 4)

Error converting payload to ControlMessage : Expecting value: line 1 column 4 (char 3)

Error converting payload to ControlMessage : Expecting value: line 1 column 3 (char 2)

Error converting payload to ControlMessage : Expecting value: line 1 column 4 (char 3)

Error converting payload to ControlMessage : Expecting property name enclosed in double quotes: line 1 column 5 (char 4)

Error converting payload to ControlMessage : Extra data: line 1 column 10 (char 9)

Error converting payload to ControlMessage : Expecting property name enclosed in double quotes: line 1 column 6 (char 5)

Error converting payload to ControlMessage : Extra data: line 1 column 10 (char 9)

Error converting payload to ControlMessage : Extra data: line 1 column 16 (char 15)

Error converting payload to ControlMessage : Extra data: line 1 column 17 (char 16)

Error converting payload to ControlMessage : Extra data: line 1 column 13 (char 12)

Error converting payload to ControlMessage : 'str' object has no attribute 'get'

Error converting payload to ControlMessage : Expecting value: line 1 column 6 (char 5)

Error converting payload to ControlMessage : Expecting value: line 1 column 4 (char 3)

Error converting payload to ControlMessage : Expecting value: line 1 column 5 (char 4)

Error converting payload to ControlMessage : Expecting property name enclosed in double quotes: line 1 column 6 (char 5)

Error converting payload to ControlMessage : Extra data: line 1 column 10 (char 9)

Error converting payload to ControlMessage : Extra data: line 1 column 16 (char 15)

Since i am trying with the example pipeline and example input, nothing changed in code and in input. Can you please suggest where am I going wrong here ? Below is the command used to run the pipeline.

![image](https://github.com/user-attachments/assets/01a67cef-9370-43c0-b2f7-61992ec3e5e4)

One more important thing is, control messages just seem to provide metadata like filepath etc but the actual messages are still read from the given filepath. Do we actually have support for reading actual data messages from Kafka instead of control messages ?

**Reply by efajardo-nv** [reply]
@srisudha-alt It appears that you are using `kafka-console-producer.sh` to send the control message to the topic as follows:
```
kafka-console-producer.sh --bootstrap-server ${BOOTSTRAP_SERVER} --topic ${MY_TOPIC} < ${FILE_TO_LOAD}
```
I believe this will send each line in the file as a message. Could you try creating a new file that has all the JSON in a single line?

**Reply by srisudha-alt** [reply, marked as answer]
@efajardo-nv  thanks for the inputs! Having Control message as a Single line json resolved the issue !

**Comment by srisudha-alt** [comment]
@efajardo-nv To have a better context of the issue we are trying to solve, I will try to explain the use case.

We want to **use Morpheus framework to perform anomaly detection** on the network traffic data. We could **run the pipelines successfully** using this example dfp_duo_pipeline.py **using input data from CSV file**. Now, we **want to use Kafka as input source** for the pipelines to do real time anomaly detection on the streaming traffic. According to the discussion we had so far, below are the takeaways we have :
1.  We couldn't use **KafkaSourceStage** in the pipeline due to input and output **type mismatch**.
2.  It has been suggested to use [dfp_integrated_training_streaming_pipeline.py](https://github.com/nv-morpheus/Morpheus/blob/branch-24.06/examples/digital_fingerprinting/production/morpheus/dfp_integrated_training_streaming_pipeline.py) instead of KafkaSourceStage. But, running this pipelines has above **issue** in previous comment. 

**Is it possible to have a 30 min sync up any time early this week to have discussion and get possible solution on this requirement ?**
